samples/nrf9160/modem_shell/src/modem_trace/modem_trace_shell.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <stdlib.h>
+#include <zephyr/shell/shell.h>
+#include <modem/nrf_modem_lib_trace.h>
+
+#if defined(CONFIG_MEMFAULT)
+#include "modem_trace_memfault.h"
+#endif
+
+#include "mosh_print.h"
+
+static uint32_t trace_start_time_ms;
+static uint32_t trace_duration_ms;
+
+int modem_traces_start(enum nrf_modem_lib_trace_level trace_level)
+{
+	int err;
+
+	mosh_print("Starting modem traces: %d", trace_level);
+
+	err = nrf_modem_lib_trace_level_set(trace_level);
+	if (err) {
+		mosh_error("Failed to enable modem traces");
+		return err;
+	}
+	trace_start_time_ms = k_uptime_get_32();
+	trace_duration_ms = 0;
+
+	return 0;
+}
+
+int modem_traces_stop(void)
+{
+	int err;
+
+	mosh_print("Stopping modem traces");
+
+	err = nrf_modem_lib_trace_level_set(NRF_MODEM_LIB_TRACE_LEVEL_OFF);
+	if (err) {
+		mosh_error("Failed to turn off modem traces");
+		return err;
+	}
+
+	trace_duration_ms = k_uptime_get_32() - trace_start_time_ms;
+	trace_start_time_ms = 0;
+
+	/* Changing the trace level to off will produce some traces, so sleep long enough to
+	 * receive those as well.
+	 */
+	k_sleep(K_SECONDS(1));
+
+	return 0;
+}
+
+void modem_traces_clear(void)
+{
+	int err;
+
+	mosh_print("Clearing modem traces");
+
+	err = nrf_modem_lib_trace_clear();
+	if (err) {
+		mosh_error("Failed to clear modem traces: %d", err);
+	}
+}
+
+void modem_trace_size(void)
+{
+	int ret;
+
+	ret = nrf_modem_lib_trace_data_size();
+	if (ret < 0) {
+		mosh_error("Failed to get modem traces size: %d", ret);
+	} else {
+		mosh_print("Modem trace data size: %d bytes", ret);
+	}
+}
+
+static int modem_trace_shell_start(const struct shell *shell, size_t argc, char **argv, void *data)
+{
+	enum nrf_modem_lib_trace_level level = (enum nrf_modem_lib_trace_level)data;
+	return modem_traces_start(level);
+}
+
+void nrf_modem_lib_trace_callback(enum nrf_modem_lib_trace_event evt)
+{
+	int err;
+
+	mosh_error("Error from nrf_modem_lib_trace: %d", evt);
+
+	if (evt == NRF_MODEM_LIB_TRACE_EVT_FULL) {
+		mosh_error("Modem trace flash area is full. Stopping modem tracing!");
+		err = modem_traces_stop();
+		if (err) {
+			mosh_error("Failed to stop modem tracing: %d", err);
+		}
+	}
+}
+
+void send_traces_to_memfault(void)
+{
+#ifdef CONFIG_MEMFAULT
+	uint32_t duration_ms = trace_duration_ms;
+
+	if (duration_ms == 0) { /* modem tracing active */
```

Comment:
It could also be that tracing has not been started? Would it be better to also check that `trace_start_time_ms` is not 0?